### ISSUE TITLE
failing to unmashall yaml with dynamic struct

### ISSUE THREAD

**Issue by sgandon** [Ompluscator/dynamic-struct#5]
hi there, 
I need to unmashall dynamically some type I don't know in advance in yaml.
I hope I could use dynamic-struct but it is failing.
here is my code
```package resources

import (
	"fmt"
	"testing"

	"github.com/davecgh/go-spew/spew"
	dynamicstruct "github.com/ompluscator/dynamic-struct"
	"gopkg.in/yaml.v3"
)

func TestReadFromYaml(t *testing.T) {
	var data = `
content-type: application/vnd.talend.iam-account
name: foo
data:
  key: value
`
	type DataResource struct {
		Key string `yaml:"key"`
	}

	type TopResource struct {
		Name        string `yaml:"name"`
		ContentType string `yaml:"content-type"`
		//		Data        DataResource `yaml:"data"`
	}

	dataResource := dynamicstruct.ExtendStruct(TopResource{}).
		AddField("Data", DataResource{}, `yaml:"data"`).
		Build().New()

	//	dataResource := TopResource{}
	if err := yaml.Unmarshal([]byte(data), &dataResource); err != nil {
		fmt.Println(err.Error())
	}
	spew.Dump(dataResource)
	reader := dynamicstruct.NewReader(dataResource)
	fmt.Println("Name", reader.GetField("Name").String())
}
```
and here is the output
```
(map[interface {}]interface {}) (len=3) {
 (string) (len=4) "name": (string) (len=3) "foo",
 (string) (len=4) "data": (map[string]interface {}) (len=1) {
  (string) (len=3) "key": (string) (len=5) "value"
 },
 (string) (len=12) "content-type": (string) (len=34) "application/vnd.talend.iam-account"
}
Failed to continue - bad access
```
First the data field is not of type DataResource which is what I expected
and then the reader fails with a bad access error.
Any idea ?

**Comment by Ompluscator** [comment]
Hi, thx for submitting an issue. After investigation, I've succeeded to have working code, by adapting a line:
``` 
// Notice, there is no "&" in front "dataResource"
if err := yaml.Unmarshal([]byte(data), dataResource); err != nil {
	fmt.Println(err.Error())
}
```
This was result in console:
```
(*struct { Name string "yaml:\"name\""; ContentType string "yaml:\"content-type\""; Data main.DataResource "yaml:\"data\"" })(0xc00008a630)({
 Name: (string) (len=3) "foo",
 ContentType: (string) (len=34) "application/vnd.talend.iam-account",
 Data: (main.DataResource) {
  Key: (string) (len=5) "value"
 }
})
Name foo
```

Package dynamic-struct relies on method reflect.New during delivering of new instance of struct, which returns pointer to new zero value. Packages "encoding/json" and "github.com/go-playground/form" still require to pass values as references while decoding/unmarshaling, but some packages like "github.com/leebenson/conform" or "gopkg.in/yaml.v3" accepts those values to be passed as non-references (as, in general, they are already created as such, from builder).

To make a general solution here would require more time for investigation, and to be honest - I'm not 100% sure that solution which supports all cases with same way of value passing for  decoding/unmarshaling can be provided (personally, I had a lot of problems dealing with combination of "github.com/go-playground/form" and "github.com/leebenson/conform").

So, I would suggest to use code example which I've added on top.
